### Feature Scaling and Train_Test Split

### What is Feature Scaling?
Feature Scaling is the process of bringing numerical features to a similar range so that features with larger values do not dominate the machine learning model.

### Normalization
Normalization rescales every feature to a fized range , typically 0 to 1 by using minimum and maximum values of that feature

Formula = x - xmin / xmax - xmin

In [2]:
import pandas as pd
import seaborn as sns

In [16]:
df = pd.DataFrame({"salary": [200,300,400,500,600]})
df

,salary
0,200
1,300
2,400
3,500
4,600


In [17]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

scaled_data = scaler.fit_transform(df)

scaled_data  #it gives array format so we use pd dataframe

minmax_used = pd.DataFrame(scaled_data, columns = df.columns)
minmax_used

,salary
0,0.00
1,0.25
2,0.50
3,0.75
4,1.00


Standardization Feature or Z_score scaling

z_score is used when there are outliers in the data , usually minmax not works well as the normall range values are compressed to zero when there are outliers.

so z_score calculate the values how far from the average better at handling outliers than normalization

In [20]:
from sklearn.preprocessing import StandardScaler

scaler2 = StandardScaler()

scaled2_data = scaler2.fit_transform(df)

z_data = pd.DataFrame(scaled2_data, columns=df.columns)

z_data

,salary
0,-1.414214
1,-0.707107
2,0.000000
3,0.707107
4,1.414214


# Test-Train Split

Divides the dataset into two subsets for machine learning

Data is divided into ratio 80:20

large training set for training the model that is 80%

small test set for testing the model that is 20%

In [52]:
from sklearn.model_selection import train_test_split

train, test = train_test_split(df, test_size = 0.2, random_state=42,shuffle = True)

train


,salary
4,600
2,400
0,200
3,500


In [53]:
test

,salary
1,300


Entire process of data preprocessing

In [54]:
df = sns.load_dataset("titanic")

In [ ]:
df

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [57]:
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

here the column deck has null 688 vallues which raises the bad impact on data 

so i am dropping the dect column

In [58]:
df = df.drop("deck",axis=1)
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


Now the age column has 177 null values

so i am filling null values using mean

In [60]:
df["age"] = df["age"].fillna(df["age"].mean())
df.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

embarked has 2 null values , filling na by using forward fill

embark_town also has 2 null values , filling na by using backward fill

In [61]:
df["embarked"] = df["embarked"].ffill()
df["embark_town"] = df["embark_town"].bfill()

df.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

Now all Null values are removed 

next step is to remove duplicated values

In [64]:
df.duplicated().value_counts()

False    780
True     111
Name: count, dtype: int64

here the 111 values repeating the same data, so we have to remove this

In [ ]:
df = df.drop_duplicates()
df.duplicated().value_counts()

False    780
Name: count, dtype: int64

In [68]:
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.000000,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.000000,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.000000,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.000000,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.000000,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
885,0,3,female,39.000000,0,5,29.1250,Q,Third,woman,False,Queenstown,no,False
887,1,1,female,19.000000,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,29.699118,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.000000,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


now , converting categorical data into numerical data using encoding techniques

for ordered categories , Label Encoder is best 

In [76]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
le = LabelEncoder()
oh = OneHotEncoder(sparse_output = False)

col = ["sex", "who", "adult_male", "alive", "alone"]

for i in col:
    df[i] = le.fit_transform(df[i])

df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,1,22.000000,1,0,7.2500,S,Third,1,1,Southampton,0,0
1,1,1,0,38.000000,1,0,71.2833,C,First,2,0,Cherbourg,1,0
2,1,3,0,26.000000,0,0,7.9250,S,Third,2,0,Southampton,1,1
3,1,1,0,35.000000,1,0,53.1000,S,First,2,0,Southampton,1,0
4,0,3,1,35.000000,0,0,8.0500,S,Third,1,1,Southampton,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
885,0,3,0,39.000000,0,5,29.1250,Q,Third,2,0,Queenstown,0,0
887,1,1,0,19.000000,0,0,30.0000,S,First,2,0,Southampton,1,1
888,0,3,0,29.699118,1,2,23.4500,S,Third,2,0,Southampton,0,0
889,1,1,1,26.000000,0,0,30.0000,C,First,1,1,Cherbourg,1,1


In [ ]:
one_hot = oh.fit_transform(df[["embark_town", "class" , "embarked"]])

one_hot = pd.DataFrame(one_hot, columns = oh.get_feature_names_out(),index=df.index)

one_hot


In [ ]:
df_new = pd.concat([df, one_hot] , axis=1)
df_new

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,...,alone,embark_town_Cherbourg,embark_town_Queenstown,embark_town_Southampton,class_First,class_Second,class_Third,embarked_C,embarked_Q,embarked_S
0,0,3,1,22.000000,1,0,7.2500,S,Third,1,...,0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
1,1,1,0,38.000000,1,0,71.2833,C,First,2,...,0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0
2,1,3,0,26.000000,0,0,7.9250,S,Third,2,...,1,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
3,1,1,0,35.000000,1,0,53.1000,S,First,2,...,0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0
4,0,3,1,35.000000,0,0,8.0500,S,Third,1,...,1,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
885,0,3,0,39.000000,0,5,29.1250,Q,Third,2,...,0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0
887,1,1,0,19.000000,0,0,30.0000,S,First,2,...,1,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0
888,0,3,0,29.699118,1,2,23.4500,S,Third,2,...,0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
889,1,1,1,26.000000,0,0,30.0000,C,First,1,...,1,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0


now remove the categorical columns

In [ ]:
df_new = df_new.drop("embarked", axis=1)
df_new = df_new.drop("embark_town", axis=1)
df_new = df_new.drop("class", axis=1)

In [85]:
df_new

,survived,pclass,sex,age,sibsp,parch,fare,who,adult_male,alive,alone,embark_town_Cherbourg,embark_town_Queenstown,embark_town_Southampton,class_First,class_Second,class_Third,embarked_C,embarked_Q,embarked_S
0,0,3,1,22.000000,1,0,7.2500,1,1,0,0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
1,1,1,0,38.000000,1,0,71.2833,2,0,1,0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0
2,1,3,0,26.000000,0,0,7.9250,2,0,1,1,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
3,1,1,0,35.000000,1,0,53.1000,2,0,1,0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0
4,0,3,1,35.000000,0,0,8.0500,1,1,0,1,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
885,0,3,0,39.000000,0,5,29.1250,2,0,0,0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0
887,1,1,0,19.000000,0,0,30.0000,2,0,1,1,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0
888,0,3,0,29.699118,1,2,23.4500,2,0,0,0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
889,1,1,1,26.000000,0,0,30.0000,1,1,1,1,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0
